### PR TITLE
Improved reporting of parser errors.

### PR DESCRIPTION
**Pull request by ryzhyk**

### CODE REVIEW

File: crates/adapters/src/controller/error.rs [hunk 1]
```diff
@@ -85,7 +92,16 @@ impl Display for ConfigError {
...
-                write!(f, "Failed to parse pipeline configuration: '{error}'")
+                write!(f, "Failed to parse pipeline configuration: {error}")
+            }
+            Self::ParserConfigParseError {
+                endpoint_name,
+                error,
+            } => {
+                write!(
+                    f,
+                    "Error parsing format configuration for endpoint '{endpoint_name}': {error}"
```
Comment (gz) [inline crates/adapters/src/controller/error.rs:103]: can we just show error here, and make endpoint_name an attribute that a user can look at?
(at least in UI context of `enpoint_name` is clear so it leads to unnecessary long error messages)

**Review by gz — APPROVED** [review]

File: crates/adapters/src/format/mod.rs [hunk 2]
```diff
@@ -1,25 +1,306 @@
...
+    /// event that caused the error, relative to the start of the stream.  In
+    /// some cases this index cannot be identified, e.g., if the error makes an entire
+    /// block of events unparseable.
+    event_number: Option<u64>,
+
+    /// Column that failed to parse.
+    ///
+    /// This field is only set when the parsing error can be attributed to a
+    /// specific column.
+    column: Option<String>,
```
Comment (mihaibudiu) [inline crates/adapters/src/format/mod.rs:140]: table column or text column?

File: crates/adapters/src/format/json/input.rs [hunk 3]
```diff
@@ -291,152 +327,280 @@ impl JsonParser {
...
-                            let json_str = String::from_utf8_lossy(bytes);
-                            anyhow!("error deserializing string as a JSON array of Debezium CDC records: {e}\nInvalid JSON: {json_str}\nExample valid JSON: '[{{\"payload\": {{\"op\": \"u\", \"before\": {{...}}, \"after\": {{...}} }} }}]'")
-                        })?;
-                        for update in updates.into_iter() {
-                            num_updates += self.apply_debezium_update(update)?;
+                Ok(update) => update,
+            };
+
+            match self.config.update_format {
+                JsonUpdateFormat::InsertDelete => {
```
Comment (mihaibudiu) [inline crates/adapters/src/format/json/input.rs:375]: should each of these cases be a separate function?

**Review by mihaibudiu — APPROVED** [review]

File: crates/adapters/src/controller/error.rs [hunk 4]
```diff
@@ -69,6 +75,7 @@ impl DetailedError for ConfigError {
     fn error_code(&self) -> Cow<'static, str> {
         match self {
             Self::ConfigParseError { .. } => Cow::from("ConfigParseError"),
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply crates/adapters/src/controller/error.rs:77]: We now have ParserConfigParserError and EncoderConfigParserError (next PR), so just FormatConfigParserError wouldn't be sufficient :)

File: crates/adapters/src/controller/error.rs (hunk 1 above)
Comment (ryzhyk) [reply crates/adapters/src/controller/error.rs:103]: This error gets thrown on pipeline startup. I suspect that right now, the UI will show the error message for the entire pipeline, without attributing it to a specific connector. If that's true, it's better to keep the endpoint name in the error message for now.

File: crates/adapters/src/controller/error.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (lalithsuresh) [reply crates/adapters/src/controller/error.rs:77]: Could we rename ConfigParseError to something more precise then? PipelineConfigParseError?

File: crates/adapters/src/format/deserializer.rs [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+            serde_json::from_str::<CaseSensitive>(r#"{"fIeLd1": true, "FIELD2": "foo"}"#)
+                .map_err(|e| e.to_string()),
+            Err(r#"missing field `field2` at line 1 column 33"#.to_string())
+        );
+    }
+
+    #[test]
+    fn error_reporting() {
+        // Correctly report parsing errors for individual fields.
+        assert_eq!(serde_json::from_str::<CaseSensitive>(r#"{"fIeLd1": 10, "field2": "foo"}"#).map_err(|e| e.to_string()), Err(r#"{"column":"fIeLd1","description":"invalid type: integer `10`, expected a boolean at line 1 column 13"} at line 1 column 13"#.to_string()));
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply crates/adapters/src/format/deserializer.rs:227]: > does the formatting tool like these long lines?

I suspect it just gives up at this point :)

File: crates/adapters/src/controller/error.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (lalithsuresh) [reply crates/adapters/src/controller/error.rs:77]: Alternatively, might even be easier to just say InvalidPipelineConfig, InvalidEncoderConfig, InvalidParserConfig etc... :)

File: crates/adapters/src/format/deserializer.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (ryzhyk) [reply crates/adapters/src/format/deserializer.rs:227]: > there's column in the text file and column in the table
> not clear from the message which one this is
> could be useful to have both if possible

I'll rename it to "field".

File: crates/adapters/src/controller/error.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ryzhyk) [reply crates/adapters/src/controller/error.rs:77]: > Could we rename ConfigParseError to something more precise then? PipelineConfigParseError?

Done

File: crates/adapters/src/format/json/input.rs (hunk 3 above)
Comment (ryzhyk) [reply crates/adapters/src/format/json/input.rs:375]: I will refactor this using traits in the next PR.

**Pull request merged**